model/src/main/java/org/projectnessie/model/Validation.java

```diff
@@ -35,17 +35,25 @@ public final class Validation {
   public static final String REF_NAME_REGEX = "^" + REF_NAME_RAW_REGEX + "$";
   public static final String REF_NAME_OR_HASH_REGEX =
       "^((" + HASH_RAW_REGEX + ")|(" + REF_NAME_RAW_REGEX + "))$";
+  public static final String REF_NAME_PATH_REGEX =
+      "^" + REF_NAME_RAW_REGEX + "(@" + HASH_RAW_REGEX + ")?$";
 
   public static final Pattern HASH_PATTERN = Pattern.compile(HASH_REGEX);
   public static final Pattern REF_NAME_PATTERN = Pattern.compile(REF_NAME_REGEX);
   public static final Pattern REF_NAME_OR_HASH_PATTERN = Pattern.compile(REF_NAME_OR_HASH_REGEX);
 
-  private static final String HASH_RULE = "consist of the hex representation of 4-32 bytes";
+  public static final String HASH_RULE = "consist of the hex representation of 4-32 bytes";
   private static final String REF_RULE =
       "start with a letter, followed by letters, digits, one of the ./_- characters, "
           + "not end with a slash or dot, not contain '..'";
 
   public static final String HASH_MESSAGE = "Hash must " + HASH_RULE;
+  public static final String REF_NAME_PATH_MESSAGE =
+      "Reference name must "
+          + REF_RULE
+          + ", optionally followed "
+          + "by ~ and a commit hash, which must "
```

Comment:
Yes, I think we wanted to use `~` I guess forgot to update the code. 

IMHO, `~` is preferable because it is not a special char in URLs. I will change to `~` unless there are objections :) WDYT?

Reply:
Either WFM

Reply:
I just released that python CLI already used the `main@11223344` syntax... I'll keep the `@` separator for now. We can change is later across all tools.